ctk/common/src/main/kotlin/org/codice/compliance/verification/core/responses/CoreAuthnRequestProtocolVerifier.kt

```diff
@@ -38,11 +42,26 @@ class CoreAuthnRequestProtocolVerifier(override val response: Node,
     }
 
     private fun verifyAuthnRequestProtocolResponse() {
-        if (response.children("Assertion")
-                        .all { it.children("AuthnStatement").isEmpty() })
-            throw SAMLComplianceException.create(SAMLCore_3_4,
+        val assertions = response.children("Assertion")
+
+        if (response.localName != "Response" || assertions.isEmpty())
+            throw SAMLComplianceException.create(SAMLCore_3_4_1_4a,
+                    message = "Did not find Response elements with one or more Assertion elements.",
+                    node = response)
+
+        if (assertions.all { it.children("AuthnStatement").isEmpty() })
+            throw SAMLComplianceException.create(SAMLCore_3_4, SAMLCore_3_4_1_4b,
                     message = "AuthnStatement not found in any of the Assertions.",
                     node = response)
+
+        if (assertions.any {
+                    it.recursiveChildren("AudienceRestriction").flatMap { it.children("Audience") }
+                            .none { it.textContent == "https://samlhost:8993/services/saml" }
```

Comment:
✏️ `SP_ISSUER` in `TestCommon` should be `https://samlhost:8993/services/saml`. I suggest using that instead.

Reply:
☝️ Definitely in favor or removing this magic string.

Reply:
Sorry, I used to do magic-oriented programming before I started going object-oriented. I slip back into that occasionally 😉